### DISCUSSION TITLE
Local development

### DISCUSSION THREAD

**Discussion by SamoKopecky** [PrivateBin/PrivateBin#1328] [Q&A] [answered]
### Describe the problem/question

Hi, I want to contribute to this project because I use it and its great! I don't want to use devcontairs as suggested in the dev README. I want to run my own IDE and my own development server for php/js. However I can't find any documentation on how to do this. In an ideal world I would run a docker container which mounts the privatebin code into it and supports hot reload so I can see the changes I make. Does something like this already exist in private or does it need to be implemented?

Thanks!

### Did you use the FAQ section?

- [X] Yes, I have read [the FAQ](https://github.com/PrivateBin/PrivateBin/wiki/FAQ) and I found no solution/answer there.

### What you did?

Doesn't apply to this question I think.

### Issue reproducibility

No, I cannot reproduce it on https://privatebin.net.

**Comment by rugk** [comment]
> ideal world I would run a docker container which mounts the privatebin code

You can [use the usual ones](https://hub.docker.com/u/privatebin) and mount stuff in there, hot reload is AFAIK not supported. But I have not used them.
What may interest you then is this unit testing container: https://hub.docker.com/r/privatebin/unit-testing

Also, BTW, if you use containers you could also use (local) devcontainers, it's basically the same. Anyway, apart from that, if you want a "traditional" experience, just use any IDE and phpunit/PHP dev tools and maybe the ones you need for JS. I can also just recommend [what the guide recommends](https://github.com/PrivateBin/PrivateBin/wiki/Development). If anything is missing, feel free to add it.

**Reply by elrido** [reply, marked as answer]
Indeed, for my local IDE-based development I too use our own container images.

To run the locally checked out repo I use the following helper script:
```shell
#!/bin/sh
docker run -d --rm --name pb -p 127.0.0.1:8080:8080 -u $(id -u) -v ~/[your/git/repo/path/here]:/var/www:rw privatebin/unit-alpine
echo "Container:  docker stop pb"
echo "Website:    http://127.0.0.1:8080/"
```

When changing JS files, you will need to update the SRI hashes (or temporarily remove them from the template used), for example by running phpunit and then `docker restart pb` to ensure the webserver- and php-op-caches get refreshed and reloading the page in the browser actually loads the new code and the container doesn't serve you stale content. Sometimes a restart is not needed (i.e. template or conf.php changes usually get picked up right away), but if you ever feel it doesn't display what you just edited, simply restart the container and refresh again.

We could offer an image without PHP opcache and webserver caching disabled or you could mount configuration to that effect into the image as well, but so far I was too lazy to explore that.

For unit testing, we offer a dedicated container image that has the AWS SDK, phpunit and all the nodejs modules incl. mocha installed in it. Here is how I launch that one:

```shell
#!/bin/sh
# script pb-unit - runs both test suites in parallel
docker run --rm --read-only -v ~/[your/git/repo/path/here]:/srv:ro privatebin/unit-testing

#!/bin/sh
# script pb-unit-js - runs only JS tests
docker run --rm --read-only -v ~/[your/git/repo/path/here]:/srv:ro privatebin/unit-testing mocha "$@"

#!/bin/sh
# script pb-unit-php - runs only PHP tests
docker run --rm --read-only -v ~/[your/git/repo/path/here]:/srv:ro privatebin/unit-testing phpunit "$@"
```

The latter two scripts pass through arguments, which is useful if you want to chase down a failure in one of the tests, without rerunning the whole thing. For example, you could call it as such: `pb-unit-js test/Model.js` or `pb-unit-php JsonApiTest.php`
